File: assemblyscript-sdk/assembly/hash.ts
// Hashing primitives — pointer-free wrappers over the §7.6 host fns.
// Mirrors the Rust reference `pyde::hash::{blake3, poseidon2, keccak256}`
// (rust/pyde-host/src/lib.rs L818): allocate the 32-byte output, call the
// host fn with (in_ptr, in_len, out_ptr), return the digest.

import { hash_blake3, hash_poseidon2, hash_keccak256 } from "@pyde-net/host/assembly";
import { Bytes32, newBytes32 } from "./types";

// Blake3 — general-purpose hashing (address derivation, event topic-0,
// content addressing).
export function blake3(input: StaticArray<u8>): Bytes32 {
  const out = newBytes32();
  hash_blake3(changetype<usize>(input), input.length, changetype<usize>(out));
  return out;
}

// Poseidon2 — Pyde's canonical storage-slot derivation
// (`slot = Poseidon2(self_address || field || key)`) and ZK-friendly
// hashing.
export function poseidon2(input: StaticArray<u8>): Bytes32 {
  const out = newBytes32();
  hash_poseidon2(changetype<usize>(input), input.length, changetype<usize>(out));
  return out;
}

// Keccak256 — EVM-interop hashing (Merkle-Patricia proofs, etc.).
export function keccak256(input: StaticArray<u8>): Bytes32 {
  const out = newBytes32();
  hash_keccak256(changetype<usize>(input), input.length, changetype<usize>(out));
  return out;
}

File: go/hash.go
//go:build tinygo

// Hashing primitives (§7.6). Each takes a byte slice and returns a
// 32-byte digest as a value — no output buffer to manage.

package pyde

// Blake3 hashes input with Blake3. General-purpose: address derivation,
// event topic-0, content addressing. Cheapest of the three.
func Blake3(input []byte) Bytes32 {
	var out Bytes32
	hashBlake3(ptr(input), int32(len(input)), ptr(out[:]))
	return out
}

// Poseidon2 hashes input with Poseidon2. ZK-friendly but more expensive
// than Blake3 in native execution; used for storage-slot derivation and
// state-root commitments. Contracts rarely need it directly — typed
// storage derives slots host-side.
func Poseidon2(input []byte) Bytes32 {
	var out Bytes32
	hashPoseidon2(ptr(input), int32(len(input)), ptr(out[:]))
	return out
}

// Keccak256 hashes input with Keccak256. Provided for cross-chain
// interop (e.g. verifying Ethereum proofs); Pyde does not use Keccak
// natively.
func Keccak256(input []byte) Bytes32 {
	var out Bytes32
	hashKeccak256(ptr(input), int32(len(input)), ptr(out[:]))
	return out
}
